### ISSUE TITLE
Bug in nuscenes_dataloader.py [os.listdir does not guarantee the order of the returned list of files on other systems beyond Ubuntu]

### ISSUE THREAD

**Issue by demmerichs** [pxiangwu/MotionNet#7]
Hi,

I think I found a bug in `data/nuscenes_dataloader.py`:
```python
            gt_file_paths = [
                os.path.join(seq_dir, f)
                for f in os.listdir(seq_dir)
                if os.path.isfile(os.path.join(seq_dir, f))
            ]
            num_gt_files = len(gt_file_paths)

            assert gt_file_paths == sorted(gt_file_paths), gt_file_paths  # <-- line inserted by me
            gt_dict_list = []
            for f in range(
                num_gt_files
            ):  # process the files, starting from 0.npy to 1.npy, etc
```
The `gt_file_paths` should contain the two files (for multi_seq training) `0.npy` and `1.npy` and based on the following comment starting the for-loop, this list should be ordered (see assertion line I inserted and marked with a comment).
Now `os.listdir` does not guarantee any ordering:
Taken from [here](https://www.tutorialspoint.com/python/os_listdir.htm):
> Python method listdir() returns a list containing the names of the entries in the directory given by path. The list is in **arbitrary** order. [...]

When I run the above code with the assertion, on the cluster everything is fine. However, on my local machine throws the assertion. So the ordering probably depends on the file system, OS, and `os.listdir` implementation, but one might get lucky and it will be always ordered.

However, I started wondering if this might even be kind of an augmentation (reverse in time), or if it even really matters (network applies to both `0.npy` and `1.npy` independently, and they are only combined for the background consistency loss, in which probably the ordering *is*(?) relevant, as one uses the `trans_matrix` which only works in one direction by definition? Could you shed some light on this, if it was intended or not?

**Comment by pxiangwu** [comment]
Hi @DavidS3141 

Yes, you are right. The listdir() has no guarantee that the returned file List is sorted. But please note that, in my code when I iterate over the files, I use this: 

```
for f in range(
    num_gt_files
): 
```
So, in this way we will iterate from 0.npy to 1.npy, thereby ensuring the order.

> this might even be kind of an augmentation (reverse in time)?

This a good point. Currently in my code I did not do any explicit data augmentation (even no flipping, rotation, scaling, etc) due to the hard disk IO issue in my system. But I think you can try this, i.e., reverse in time.

> or if it even really matters ... in which probably the ordering is(?) relevant,

Currently in my code, I think the ordering matters since I expect my code to run 0.npy first and then 1.npy.

**Comment by demmerichs** [comment]
> But please note that, in my code when I iterate over the files, I use this:
```python
for f in range(
    num_gt_files
):  # process the files, starting from 0.npy to 1.npy, etc
    gt_file_path = gt_file_paths[f]
    ...
```

> or if it even really matters ... in which probably the ordering is(?) relevant,

>> Currently in my code, I think the ordering matters since I expect my code to run 0.npy first and then 1.npy.

Just realized your argument is not correct, I extended the code snippet which shows, that you just iterate over the filenames in the order `os.listdir` returned them. Your for-loop is the same as:
```python
for gt_file_path in gt_file_paths:  # process the files, starting from 0.npy to 1.npy, etc
    ...
```

**Comment by pxiangwu** [comment]
Hi @DavidS3141 

Hmmm ... This is interesting. I remember that I used the code like this:

```
for f in range(
    num_gt_files
):  # process the files, starting from 0.npy to 1.npy, etc
    gt_file_path = str(f) + ".npy"
    ....
```

I checked the current uploaded code at MERL and it seems like it cannot guarantee the file order as you mentioned. And I test `os.listdir` on Ubuntu and it returns 0.npy and 1.npy in order. But this may not be guaranteed in other systems such as Windows or Unix, etc.

And yes, to guarantee the order, one can either sort the file list or use the code snippet above.

Thanks a lot for pointing out this.
